### PR TITLE
Skeleton app

### PR DESCRIPTION
**Pull request by nikkolasg**
This is the basic code to have when one wants to implement a new app using cothority trees. Note this app is to use with deploy/ lib, not as a standalone app. In this way, we can easily deploy a new app and make measurements and have nice graphs !

### CODE REVIEW

File: app/skeleton/skeleton.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/dedis/cothority/lib/conode"
+	"github.com/dedis/cothority/lib/dbg"
+	"github.com/dedis/cothority/lib/monitor"
+	"github.com/dedis/cothority/lib/sign"
+	"time"
+)
+
+// This file is the first draft to a skeleton app where you have all the
+// basics to run your own cothority tree. This include the main where you handle
+// the configuration + the "running" part. It also include a basic Round
```
Comment (Liamsi) [inline app/skeleton/skeleton.go]: Maybe better?
This includes a main() function which shows how to configure and run a cothority based application.

File: app/skeleton/skeleton.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+// current version of the API and a lot of changes will be brought along the
+// next months, so of course there's a lot of things that are not ideal, we know
+// that ;).
+
+// To run this skeleton app, go to deploy:
+// go build && ./deploy -debug 2 simulations/skeleton.toml
+
+func main() {
+	// First, let's read our config
+	// YOu should create your own config in lib/app.
```
Comment (Liamsi) [inline app/skeleton/skeleton.go]: typo

File: app/skeleton/skeleton.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	hostname := app.RunFlags.Hostname
+	// i.e. we are root
+	if hostname == conf.Hosts[0] {
+		dbg.Lvlf3("Tree is %+v", conf.Tree)
+	}
+	dbg.Lvl3(hostname, "Starting to run")
+
+	// Connect to the monitor process. This monitor process is run on your
+	// machine and accepts connections from any node, usually you only connect
+	// with the root for readability and performance reason (don't connect to
```
Comment (Liamsi) [inline app/skeleton/skeleton.go]: reasons

File: lib/conode/peer.go [hunk 4]
```diff
@@ -54,7 +54,7 @@ func NewPeer(address string, conf *app.ConfigConode) *Peer {
 	dbg.Lvl3("loading configuration")
 	var hc *graphs.HostConfig
 	opts := graphs.ConfigOptions{ConnType: "tcp", Host: address, Suite: suite}
-
+	dbg.Print("Config before loadconfig address:", address, "conf:", conf)
```
Comment (Liamsi) [inline lib/conode/peer.go]: Is this `dbg.Print` intended? Almost everywhere else we use dbg.Lvl* instead

File: app/skeleton/skeleton.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	sign.RegisterRoundFactory(RoundSkeletonType,
+		func(node *sign.Node) sign.Round {
+			return NewRoundSkeleton(node)
+		})
+
+	// Here it will create a new round each seconds automatically.
+	// If you need more fined grained control, you must implement yourself the
+	// conode.Peer struct (it's quite easy).
+	peer.LoopRounds(RoundSkeletonType, conf.Rounds)
+	// Notify we finished to monitor so that the simulation can be stopped
```
Comment (Liamsi) [inline app/skeleton/skeleton.go]: (maybe?) Notify the monitor that we finished so ...

File: app/skeleton/skeleton.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	dbg.Lvl3("Making new RoundSkeleton", node.Name())
+	round := &RoundSkeleton{}
+	// You've got to initialize the roundcosi with the node
+	round.RoundCosi = sign.NewRoundCosi(node)
+	round.Type = RoundSkeletonType
+	return round
+}
+
+// The first phase is the announcement phase.
+// For all phases, the signature is the same, it takes sone Input message and
```
Comment (Liamsi) [inline app/skeleton/skeleton.go]: typo: some

**Comment by Liamsi** [comment]
I think it's OK to merge. My comments are just nitpicking as it is a first draft and the skeleton is for 'internal use' only (but very useful, thanks)

**Comment by ineiti** [comment]
Consider merging #138, then I'm OK to merge this to development.

**Comment by nikkolasg** [comment]
Then I merge this one then #138

**Pull request merged**